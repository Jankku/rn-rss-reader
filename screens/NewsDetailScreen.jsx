import { useTheme } from '@react-navigation/native';
import { useEffect, useState } from 'react';
import { View } from 'react-native';
import WebView from 'react-native-webview';
import ArticleSaveButton from '../components/NewsDetail/ArticleSaveButton';
import FeedController from '../data/FeedController';
import SavedArticleController from '../data/local/SavedArticleController';
import useToast from '../hooks/useToast';

function NewsDetailScreen({ navigation, route }) {
  const guid = route.params?.guid;
  const { colors } = useTheme();
  const { showToast } = useToast();
  const [article, setArticle] = useState();
  const [isSaved, setIsSaved] = useState();
  const openWebView = article && article.encoded === undefined;
  const bodyFontSize = '2.25rem';
  const figureFontSize = '2.0rem';

  useEffect(() => {
    const item = FeedController.findItemById(guid);
    setArticle(item);

    const saved = SavedArticleController.isArticleSaved(guid);
    setIsSaved(saved);
  }, [guid]);

  useEffect(() => {
    if (article) {
      navigation.setOptions({
        title: article.title,
        headerTitleAlign: 'left',
        headerRight: () => (
          <ArticleSaveButton
            color={colors.headerText}
            isSaved={isSaved}
            onPress={() => {
              if (article && isSaved) {
                deleteArticleAction();
              } else {
                saveArticleAction();
              }
            }}
          />
        ),
      });
    }
  }, [article, isSaved]);

  const saveArticleAction = () => {
    try {
      const dbArticle = SavedArticleController.articleToDbArticle(article);
      SavedArticleController.saveArticle(dbArticle);
      setIsSaved(true);
      showToast('Article saved');
    } catch (error) {
      showToast("Error: Couldn't save article");
      console.error(error);
    }
  };

  const deleteArticleAction = () => {
    try {
      const dbArticle = SavedArticleController.getArticleById(guid);
      SavedArticleController.deleteArticle(dbArticle);
      setIsSaved(false);
      showToast('Article deleted');
    } catch (error) {
      showToast("Error: Couldn't delete article");
      console.error(error);
    }
  };

  const injectedJS = `
    if (window.location.href === 'about:blank') {
      document.body.style.color = "${colors.text}";
      document.body.style.fontSize = "${bodyFontSize}";
      document.body.style.padding = "8px";

      const figures = document.querySelectorAll("figure");
      figures.forEach((f) => {
        f.style.margin = 0;
        f.style.fontSize = "${figureFontSize}";
        f.style.fontStyle = "italic";
      });

      const images = document.querySelectorAll("img");
      images.forEach((i) => {
        i.style.marginRight = "16px";
      });

      const links = document.querySelectorAll("a");
      links.forEach((l) => {
        l.style.color = "${colors.text}";
      });
    }

    true;
    `;

  return (
    <View style={{ flex: 1 }}>
      {openWebView ? (
        <WebView
          source={{
            uri: article.link,
          }}
        />
      ) : article?.encoded ? (
        <WebView
          injectedJavaScriptForMainFrameOnly
          scalesPageToFit
          injectedJavaScript={injectedJS}
          originWhitelist={['*']}
          source={{
            html: article.encoded,
          }}
          style={{
            backgroundColor: colors.card,
          }}
        />
      ) : null}
    </View>
  );
}

export default NewsDetailScreen;
