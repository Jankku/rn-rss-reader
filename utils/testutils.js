import { render } from '@testing-library/react-native';
import { ThemeContextProvider } from '../context/ThemeContext';
import { LocationContextProvider } from '../context/LocationContext';
import { RegionContextProvider } from '../context/RegionContext';

const AllProviders = ({ children }) => {
  return (
    <LocationContextProvider>
      <RegionContextProvider>
        <ThemeContextProvider>{children}</ThemeContextProvider>
      </RegionContextProvider>
    </LocationContextProvider>
  );
};

const customRender = (ui, options) => render(ui, { wrapper: AllProviders, ...options });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };
