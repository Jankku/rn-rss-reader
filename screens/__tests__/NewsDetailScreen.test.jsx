import NewsDetailScreen from '../NewsDetailScreen';
import { render } from 'testutils';

const articleData = {
  category: ['Category1', 'Category2', 'Category3'],
  description: 'This is a description.',
  encoded: 'Test',
  guid: {
    '#text': 'https://example.com',
    isPermaLink: 'false',
  },
  link: 'https://example.com',
  pubDate: 'Fri, 08 Apr 2022 09:19:43 +0300',
  title: 'This is a title.',
};

jest.mock('../../data/FeedController', () => {
  return {
    findItemById: jest.fn().mockImplementation(() => articleData),
  };
});

describe('Test NewsDetailScreen', () => {
  test('Should match snapshot', () => {
    const { toJSON } = render(
      <NewsDetailScreen
        navigation={{ navigate: () => {}, setOptions: () => {} }}
        route={{ params: { guid: 'https://yle.fi/uutiset/3-12345678' } }}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
