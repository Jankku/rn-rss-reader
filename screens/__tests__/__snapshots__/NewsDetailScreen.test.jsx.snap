// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NewsDetailScreen Should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    animated={true}
    animationConfig={
      Object {
        "bounciness": 0,
      }
    }
    animationType="spring"
    indeterminate={false}
    indeterminateAnimationDuration={1000}
    onLayout={[Function]}
    progress={0}
    style={
      Array [
        Object {
          "backgroundColor": "#f7f7f7",
          "borderColor": "#0a3b77",
          "borderRadius": 0,
          "borderWidth": 0,
          "overflow": "hidden",
          "width": 750,
        },
        undefined,
      ]
    }
    useNativeDriver={true}
  >
    <View
      style={
        Object {
          "backgroundColor": "#0a3b77",
          "height": 3,
          "transform": Array [
            Object {
              "translateX": -2.842170943040401e-14,
            },
            Object {
              "translateX": -375,
            },
            Object {
              "scaleX": 0.0001,
            },
          ],
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <RNCWebView
      cacheEnabled={true}
      injectedJavaScript="
    if (window.location.href === 'about:blank') {
      document.body.style.color = \\"#101010\\";
      document.body.style.fontSize = \\"2.25rem\\";
      document.body.style.padding = \\"8px\\";

      const figures = document.querySelectorAll(\\"figure\\");
      figures.forEach((f) => {
        f.style.margin = 0;
        f.style.fontSize = \\"2.0rem\\";
        f.style.fontStyle = \\"italic\\";
      });

      const images = document.querySelectorAll(\\"img\\");
      images.forEach((i) => {
        i.style.marginRight = \\"16px\\";
      });

      const links = document.querySelectorAll(\\"a\\");
      links.forEach((l) => {
        l.style.color = \\"#101010\\";
      });
    }

    true;
    "
      injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
      injectedJavaScriptForMainFrameOnly={true}
      javaScriptEnabled={true}
      messagingEnabled={false}
      onContentProcessDidTerminate={[Function]}
      onHttpError={[Function]}
      onLayout={[Function]}
      onLoadingError={[Function]}
      onLoadingFinish={[Function]}
      onLoadingProgress={[Function]}
      onLoadingStart={[Function]}
      onMessage={[Function]}
      onScroll={[Function]}
      onShouldStartLoadWithRequest={[Function]}
      scalesPageToFit={true}
      source={
        Object {
          "html": "Test",
        }
      }
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#ffffff",
          },
          Object {
            "backgroundColor": "#f7f7f7",
          },
        ]
      }
      useSharedProcessPool={true}
    />
  </View>
</View>
`;
